components/sup/src/manager/service/health.rs

```diff
@@ -26,12 +46,167 @@ impl From<i8> for HealthCheck {
-        };
-        write!(f, "{}", msg)
+        match *self {
+            HealthCheck::Ok => "OK".fmt(f),
+            HealthCheck::Warning => "WARNING".fmt(f),
+            HealthCheck::Critical => "CRITICAL".fmt(f),
+            HealthCheck::Unknown => "UNKNOWN".fmt(f),
+        }
+    }
+}
+
+/// All state needed for checking the health of a service over time.
+#[derive(Clone)]
+pub struct State {
+    // All hooks currently need these
+    hook:                   Option<Arc<HealthCheckHook>>,
+    service_group:          ServiceGroup,
+    package:                Pkg,
+    svc_encrypted_password: Option<String>,
+
+    /// A reference to the process supervisor for the service. This is
+    /// used to create a "proxy health check" for services that do not
+    /// provide their own health check hook.
+    supervisor: Arc<Mutex<Supervisor>>,
+
+    /// The configured interval at which to run health checks for this
+    /// service. The interval actually used may differ based on the
+    /// previous health check result.
+    nominal_interval: HealthCheckInterval,
+
+    /// A reference to the service's current health check status. We
+    /// store the result of the health check here.
+    service_health_result: Arc<Mutex<HealthCheck>>,
+
+    /// A reference to the Supervisor's gateway state. We also store
+    /// the status in here for making it available via the HTTP
+    /// gateway.
+    gateway_state: Arc<RwLock<GatewayState>>,
+}
+
+impl State {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(hook: Option<Arc<HealthCheckHook>>,
+               service_group: ServiceGroup,
+               package: Pkg,
+               svc_encrypted_password: Option<String>,
+               supervisor: Arc<Mutex<Supervisor>>,
+               nominal_interval: HealthCheckInterval,
+               service_health_result: Arc<Mutex<HealthCheck>>,
+               gateway_state: Arc<RwLock<GatewayState>>)
+               -> Self {
+        State { hook,
+                service_group,
+                package,
+                svc_encrypted_password,
+                supervisor,
+                nominal_interval,
+                service_health_result,
+                gateway_state }
+    }
+
+    /// Creates a future that runs the health check and then waits for
+    /// a suitable interval. Multiple such iterations will then be
+    /// chained together for an unending stream of health checks.
+    fn single_iteration(self) -> impl Future<Item = (), Error = ()> {
+        let State { hook,
+                    service_group,
+                    package,
+                    svc_encrypted_password,
+                    supervisor,
+                    nominal_interval,
+                    service_health_result,
+                    gateway_state, } = self;
+
+        // Use an Arc to avoid having to have full clones everywhere. :/
+        let service_group = Arc::new(service_group);
+        let service_group_copy = service_group.clone();
```

Comment:
```suggestion
        let service_group_ref = Arc::clone(&service_group);
```
Use [the preferred idiom](http://doc.rust-lang.org/1.34.1/std/sync/struct.Arc.html#cloning-references), plus I think `_copy` is misleading since we're explicitly trying to avoid a copy.